scripts/user_data_pubcloud.sh

```diff
@@ -1,23 +1,104 @@
-
-source /etc/lsb-release
-
-# Delete configuration which enables automatic upgrades
-# to prevent the instance packages being upgraded before
-# the correct apt repositories have been configured.
-# ref: RE-458 / RE-473
-if [[ "${DISTRIB_CODENAME}" == "trusty" ]]; then
-    # The 'unattended-upgrades' package cannot
-    # be purged on trusty as it is a dependency
-    # for cloud-init. So instead we just remove
-    # the configuration.
-    rm -f /etc/apt/apt.conf.d/*unattended-upgrades
+#!/bin/bash -e
+
+# This script is used as a user data file when building
+# machines which use images that are not managed by
+# nodepool. See https://cloud-init.io for more information
+# about how cloud-init uses user-data files.
+#
+# This script tries to ensure that the machine is prepared
+# in a consistent way and it has all the required packages
+# installed on it for jenkins/ansible to access it. Any
+# further preparation should be done using jenkins/ansible.
+
+# Get the distribution name
+echo -e "\n### Getting the distribution name ###\n"
+if [[ -e /etc/lsb-release ]]; then
+  source /etc/lsb-release
+  DISTRO_RELEASE=${DISTRIB_CODENAME}
+elif [[ -e /etc/os-release ]]; then
+  source /etc/os-release
+  DISTRO_RELEASE=${UBUNTU_CODENAME}
 else
-    export DEBIAN_FRONTEND=noninteractive
-    apt-get purge -y unattended-upgrades
+  echo "Unable to determine distribution due to missing lsb/os-release files."
+  exit 1
 fi
+
+# Save a backup of the original apt sources
+echo -e "\n### Saving a backup of the original apt sources ###\n"
+mv /etc/apt/sources.list /etc/apt/sources.list.original
+
+# Rewrite the apt sources file
+echo -e "\n###Rewriting the apt sources ###\n"
+DISTRO_MIRROR="http://mirror.rackspace.com/ubuntu"
+DISTRO_COMPONENTS="main,universe"
+cat << EOF >/etc/apt/sources.list
+deb ${DISTRO_MIRROR} ${DISTRO_RELEASE} ${DISTRO_COMPONENTS//,/ }
+deb ${DISTRO_MIRROR} ${DISTRO_RELEASE}-updates ${DISTRO_COMPONENTS//,/ }
+deb ${DISTRO_MIRROR} ${DISTRO_RELEASE}-backports ${DISTRO_COMPONENTS//,/ }
+deb ${DISTRO_MIRROR} ${DISTRO_RELEASE}-security ${DISTRO_COMPONENTS//,/ }
+EOF
+
+# Enable debug logging for apt to make diagnosis of apt failures easier
+echo -e "\n### Enabling apt debug logging ###\n"
+echo 'Debug::Acquire::http "true";' > /etc/apt/apt.conf.d/99debug
+
+# Add jenkins user and group
+echo -e "\n### Adding the jenkins user and group ###\n"
+JENKINS_HOME="/var/lib/jenkins"
+groupadd jenkins
+useradd --gid jenkins \
+        --shell /bin/bash \
+        --home-dir ${JENKINS_HOME} \
+        --create-home jenkins
+
+# Fetch the rcbops public keys and add them to authorized_keys
+echo -e "\n### Fetching the rcbops public keys ###\n"
+SSH_PUBLIC_KEYS_URL="https://raw.githubusercontent.com/rcbops/rpc-gating/master/keys/rcb.keys"
+curl --connect-timeout 5 \
+     --retry 3 \
+     ${SSH_PUBLIC_KEYS_URL} > /tmp/ssh-public-keys
+
+echo -e "\n### Configuring authorized_keys for jenkins and root ###\n"
+for usr_home in /root ${JENKINS_HOME}; do
```

Comment:
If the keys are already there, what happens? If there are duplicates in authorized_keys, what happens?

Reply:
It sill works if there are duplicates.